### DISCUSSION TITLE
Support for `bibtex` files

### DISCUSSION THREAD

**Discussion by Cs137** [saulpw/visidata#2180] [Ideas]
Hi and many thanks for this awesome piece of software! :heart_eyes: 

I would like to use vd to manipulate my bibtex library. I can load a [bibtex](https://www.bibtex.org/Format/) file into vd, but it would be very convenient if the data would be tabulated, like e.g. json data. The bibtex-key + all existing tags could be presented as column headers with rows for the individual entries. Is there any way to achieve this currently already? Glad for any advice

**Comment by saulpw** [comment]
Hi, @Cs137, thanks and you're welcome!  Tell me, how would you extract the data in Python?  If you can get a list of dicts (or other Python object), we can make a loader for it really easily!

**Reply by Cs137** [reply]
Many thanks for your fast response.

I have the snipped underneath, It returns the entries as list of dicts, with a `type` and a `fields` key. The "type" value is a string, and the fields value is a dict with the actual entrie's data. It is for sure not the most efficient way to obtain the data, but it worked for me so far ([example bibtex file](http://tug.ctan.org/tex-archive/macros/latex/contrib/aguplus/sample.bib)).

```python
def read_bibtex(file_path: str, encoding: str='utf-8') -> list[dict[str,any]]:
    with open(file_path, 'r', encoding=encoding) as file:
        entries = []
        current_entry = None
        for line in file:
            line = line.strip()
            if not line or line.startswith('%'):
                # Skip empty lines and comments
                continue
            if line.startswith('@'):
                # New entry
                if current_entry:
                    entries.append(current_entry)
                current_entry = {'type': line.split('{')[0][1:], 'fields': {}}
            elif line.startswith('}'):
                # End of entry
                if current_entry:
                    entries.append(current_entry)
                current_entry = None
            else:
                # Field
                parts = line.split('=')
                if len(parts) == 2:
                    key = parts[0].strip().lower()
                    value = parts[1].strip().strip('{}"')
                    if current_entry:
                        current_entry['fields'][key] = value[:-2]
        return entries
```

**Reply by saulpw** [reply]
Okay, change it to this, and put in your .visidatarc:

```
@VisiData.api
def open_bibtex(vd, p):
    return BibtexSheet(p.name, source=p)

class BibtexSheet(Sheet):
    columns = [
        ItemColumn('type'),
        ItemColumn('fields.author'),
        ItemColumn('fields'),
    ]
    def iterload(self):
        for line in self.source.open_text():
            ... your code here ...
            yield current_entry
```

If that works, then you can add the imports and submit a PR to create a loader for it!

**Reply by Cs137** [reply]
Nice, many many thanks! 

I will try to find some time after my sleeping period (EU)  for it and will modify the loader to obtain one sheet, containing the columns with items for both, the type and the actual data. And as add-on a nested author column, which requires some adjustment in the loader as well.

**Reply by Cs137** [reply]
Dear @saulpw,

I continued looking into the bibtex integration, but am struggling with basics trying to set up the loader.

I created a `~/visidata/config.py` file with the content suggested by you, but have the impression the `iterload` method is not called at all. I added some print statements and opened a bib file with with vd using the `--debug` flag, the last print statement I get is the assignment of the columns:

```python
COL_ENTRYTYPE = 'type'
COL_CITATIONKEY = 'citationkey'
COL_FIELDS = 'fields'

@VisiData.api
def open_bibtex(vd, p):
    print("Started 'open_bibtex' function...")         # <-- This is not printed
    return BibtexSheet(p.name, source=p)


class BibtexSheet(Sheet):
    print("Initialised 'BibtexSheet' instance...")     # <-- This is printed
    columns = [
        ItemColumn(COL_ENTRYTYPE),
        ItemColumn(COL_CITATIONKEY),
        ItemColumn(COL_FIELDS),
        ItemColumn(COL_FIELDS + '.author'),
    ]
    print("Columns: ", columns)                        # <-- This is printed

    def iterload(self):
        print("Started 'iterload' method...")          # <-- This is not printed
        current_entry = None
        for line in self.source.open_text():
            line = line.strip()
            if not line or line.startswith("%"):
                continue
            if line.startswith("@"):
                if current_entry:
                    entries.append(current_entry)
                parts = line.split("{")
                current_entry = {
                    COL_ENTRYTYPE: parts[0][1:].lower(),
                    COL_CITATIONKEY: parts[1][:-1],
                    COL_FIELDS: {},
                }
            elif line.startswith("}"):
                if current_entry:
                    entries.append(current_entry)
                current_entry = None
            else:
                parts = line.split("=")
                if len(parts) == 2:
                    key = parts[0].strip().lower()
                    value = parts[1].strip()[:-1].strip('{}"')
                    if current_entry:
                        current_entry[COL_FIELDS][key] = value
            if current_entry is not None:
                if ' and ' in current_entry[COL_FIELDS]['author']:
                    authors = current_entry[COL_FIELDS]['author'].split(' and ')
                else:
                    authors = [current_entry[COL_FIELDS]['author']]
                current_entry[COL_FIELDS]['author'] = authors

                yield current_entry
```

I created in the meantime also a [small python package](https://codeberg.org/Cs137/bib2tex), which does what I had in mind to do with vd; but out of curiosity I'd like to get it implemented in vd as well. 

The [parser of my package](https://codeberg.org/Cs137/bib2tex/src/commit/da040a3d81d4907f9df044e4abba96a6b5ef70cd/bib2tex/bibtex_parser.py#L88) handles the author assignment better than the example above and tries to get the first and last names for the individual authors from the string, which are added to a list of dicts. The latter replaces the content of the author field. 

The resulting entries list can also be used to initialise a pandas df, maybe returning that df as Sheet might be a better approach for an integration of bibtex into vd. But I have to admit that I am a bit lost in how to adjust the loader, even though I looked into your [api docs](https://www.visidata.org/docs/api/loaders). Would you be so kind to give me another hint?

Best regards,
Christian

**Reply by saulpw** [reply]
Hi @Cs137, your `print` statements are running unconditionally at import time, and not at runtime; only the ones in `open_bibtex` and `iterload` will actually run because it's trying to open a .bibtex file.  You should use `vd.status()` for these anyway.

The code looks fine at a first glance.  I'm not actually sure why it's importing the file at `~/visidata/config.py` though.  The actual file you should put these in, until you're more familiar with how Python imports work in general, is `~/.visidatarc`.

Please include a test bibtex file, and the actual command-line you use to open it, so we can try to reproduce your issue, and figure out what's going on.

**Reply by Cs137** [reply]
Hi @saulpw, thanks for your fast response and sorry for my mistake. The file's path is `~/.config/visidata/config.py`. I was first unsure whether it gets imported at all, that's why I started adding the print statements. The bib file I am trying to open is the one linked above, I added a wget command to obtain the file and the way how I try opening it underneath:

```bash
wget http://tug.ctan.org/tex-archive/macros/latex/contrib/aguplus/sample.bib
vd --debug --open_bibtex sample.bib
```

I added the `--open_bibtex` just as a test. It is also not parsed without it. I think an assignment that the bib file should be handled by the `open_bibtex` function is missing somewhere.

**Reply by Cs137** [reply]
Sorry for my unknowledge, `vd --debug -f bibtex sample.bib` does the job and shows that there is some work on the parser left. I will pick it up tomorrow and will find my way to add the loader to vd based on your docs. Thanks for your help!

**Reply by saulpw** [reply]
Ah yes, and if `.bib` is the common extension for these files, just name the function `open_bib`.
